### PR TITLE
Basic structure of world parser and builder

### PR DESCRIPTION
**Pull request by JetteSchumann**
Finally a draft for the worldparser (python) and WorldBuilder (cpp). ~~Main work is done in the WorldBuilder so that the python file looks pretty empty now.~~

We decided for:
- fixed names of dxf laxers: LevelX, LevelX_specialAreas -> assignment of level and its areas is done by name
- store positions ~~in mm as int~~ as Length Unit
- process only lines of dxf files (no circles, ...)
- check for valid special areas in WorldBuilder
- unique color/id for each area (also for special areas of same type)

There are still lots of todos:
- [x] pybind
- [ ] ~~exception handling~~: makes sense when more functionalities are implemented at a later point
- [ ] ~~functionality of adapting dimensions of the world~~ (especially needed for reading in SUMO geo data since position data might exceed the valid int range): we might need this later but not for our first simulation
- [ ] ~~logging~~: makes sense when more functionalities are implemented at a later point
- [ ] ~~testing~~: new issue
- [x] code documentation
- [x] use Length Unit
...

### CODE REVIEW

File: jps/jupedsim/IO/worldparser.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        Method initiates parsing of Lines and Areas for each level provided in the dxf file
+        :return:
+        """
+
+        # TODO: loop over several levels, error handling
+        self.__parseLineSegment("Level0")
+        self.__parseSpecialAreas("Level0_SpecialAreas")
+        # TODO: parse polylines
+
+    def __parseLineSegment(self, p_layer: str):
```
Comment (zeroset) [inline jps/jupedsim/IO/worldparser.py:79]:
```suggestion
    def __parseLineSegment(self, p_layer: str) -> None:
```

File: jps/jupedsim/IO/worldparser.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                    ),
+                    jpscore.geometry.Coordinate(
+                        jpscore.geometry.LengthUnit(coords[2]),
+                        jpscore.geometry.LengthUnit(coords[3]),
+                        jpscore.geometry.Level(),
+                    ),
+                ),
+            )
+
+    def __parseSpecialAreas(self, p_layer: str):
```
Comment (zeroset) [inline jps/jupedsim/IO/worldparser.py:103]:
```suggestion
    def __parseSpecialAreas(self, p_layer: str) -> None:
```

File: jps/jupedsim/IO/worldparser.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
+import re
+from typing import List
+
+import ezdxf
+import jpscore
```
Comment (zeroset) [inline jps/jupedsim/IO/worldparser.py:5]: I would suggest only to import geometry sub-module
```suggestion
from jpscore import geometry
```

File: jps/jupedsim/IO/worldparser.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        # build World
+        self.m_jps_world_builder = jpscore.geometry.WorldBuilder()
+        self.__parseLevels()
+
+    @staticmethod
+    def parseCoordinates(line: str) -> List[float]:
+        """
+        Method parses a dxf line with coordinates to a list
+        :param line: start and end coordinate of a dxf line formatted as string '(start_x, start_y, start_z) (end_x, end_x, end_z)'
+        :return: coordinates as a list [start_x, start_y, end_x, end_y]
```
Comment (zeroset) [inline jps/jupedsim/IO/worldparser.py:57]: Maybe something like: 
```suggestion
        :return: coordinates as a list [{"x": x, "y": y}, {"x": x, "y": y}]
```
But i am not sure if this any better.

File: jps/jupedsim/IO/worldparser.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        """
+        for line in self.m_msp.query('LINE[layer=="' + p_layer + '"]'):
+            coords = WorldParser.parseCoordinates(line)
+            self.m_jps_world_builder.addLineSegment(
+                jpscore.geometry.Level(),
+                jpscore.geometry.LineSegment(
+                    jpscore.geometry.Coordinate(
+                        jpscore.geometry.LengthUnit(coords[0]),
+                        jpscore.geometry.LengthUnit(coords[1]),
+                        jpscore.geometry.Level(),
```
Comment (zeroset) [inline jps/jupedsim/IO/worldparser.py:93]: The Level should be parsed from the Layername and set correspondingly.

File: cpp/libcore/source/geometry/world_builder.hpp [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+class WorldBuilder
+{
+private:
+    std::map<jps::Level, std::vector<jps::LineSegment>> m_level_to_line_segments;
+    std::map<jps::Level, std::vector<jps::SpecialArea>> m_level_to_special_areas;
+
+public:
+    WorldBuilder() = default;
+
+    auto addLineSegment(const jps::Level & p_level, const jps::LineSegment & p_segment) -> void;
```
[the remark this answers is not in the document]
Comment (JetteSchumann) [reply cpp/libcore/source/geometry/world_builder.hpp:21]: We decided to store the information per level. p_level is redundant here and the same as in the coordinates of the segment. We decided to keep it for better readability.

File: cpp/libcore/source/geometry/world_builder.cpp [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#include "world_builder.hpp"
+
+auto jps::WorldBuilder::addLineSegment(
+    const jps::Level & p_level,
+    const jps::LineSegment & p_segment) -> void
+{
+    // TODO: ensure that coordinates belong to same level, add to map
```
[the remark this answers is not in the document]
Comment (JetteSchumann) [reply cpp/libcore/source/geometry/world_builder.cpp:7]: Changed the comment

File: jps/jupedsim/IO/worldparser.py (hunk 4 above)
Comment (JetteSchumann) [reply jps/jupedsim/IO/worldparser.py:57]: The function returns a List of geometry.Coordinates now

File: jps/jupedsim/IO/worldparser.py (hunk 5 above)
Comment (JetteSchumann) [reply jps/jupedsim/IO/worldparser.py:93]: Level ist created in __parseLevels() now. Created a todo for parsing

**Comment by JetteSchumann** [comment]
Next TODOs are listed in #48

File: cpp/libcore/source/geometry/world_builder.hpp [hunk 8]
```diff
@@ -0,0 +1,26 @@
+#ifndef SIMULATOR_WORLD_BUILDER_HPP
+#define SIMULATOR_WORLD_BUILDER_HPP
```
Comment (schroedtert) [inline cpp/libcore/source/geometry/world_builder.hpp:2]:
```
#pragma once
```

File: cpp/libcore/source/geometry/world_builder.hpp [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+public:
+    WorldBuilder() = default;
+
+    auto addLineSegment(const jps::Level & p_level, const jps::LineSegment & p_segment) -> void;
+    auto addSpecialArea(const jps::Level & p_level, const jps::SpecialArea & p_area) -> void;
+};
+} // namespace jps
+
+#endif // SIMULATOR_WORLD_BUILDER_HPP
```
Comment (schroedtert) [inline cpp/libcore/source/geometry/world_builder.hpp:26]: delete (not needed with `#pragma once`

File: cpp/libcore/source/geometry/world_builder.hpp (hunk 8 above)
Comment (JetteSchumann) [reply cpp/libcore/source/geometry/world_builder.hpp]: thanks

File: cpp/libcore/source/geometry/world_builder.hpp (hunk 9 above)
Comment (JetteSchumann) [reply cpp/libcore/source/geometry/world_builder.hpp]: Done

**Review by zeroset — APPROVED** [review]
LGTM

**Pull request merged**